### ISSUE TITLE
[Question] VarStore and Path

### ISSUE THREAD

**Issue by tadas-subonis** [LaurentMazare/tch-rs#167]
What's the purpose of having VarStore and Path? I've used PyTorch python and briefly looked at C++ version but I've never seen this concept. Also, the documentation is not entirely clear about that.

Why is it good/necessary to have them?

Thanks!

**Comment by LaurentMazare** [comment]
VarStore are used to store variables and paths to group them. E.g. if you look at the [resnet example](https://github.com/LaurentMazare/tch-rs/blob/master/src/vision/resnet.rs), variables are placed in the 'layer1', 'layer2', 'layer3', 'layer4' top paths, then for each of these path we have subpath based on the block index in the layer and so on. This makes the full name of the variables a bit more explicit.

When using TensorFlow or PyTorch in Python, you would by default get some implicit names as well as an implicit variable store that contains all variables, and you would able to override the paths with `tf.name_scope` etc. Here we have a more explicit approach, the main advantage to me is that it's less confusing than some implicit naming scheme.

Finally VarStore can be used to load/store trained weights, and are passed to the optimizer so that it knows which variables to optimize.

**Comment by tadas-subonis** [comment]
Ok. Thanks. So in the end, this basically helps with debugging?

**Comment by LaurentMazare** [comment]
Not specifically debugging, I would say that it helps knowing what the variable names/paths are (rather than having this being automatically defined at runtime, with potential issues e.g. if the model is not built in exactly the same way when you load some variables)

**Comment by tadas-subonis** [comment]
Awesome. Thanks
